### ISSUE TITLE
Loss became non sometimes potentially with SpecAugment 

### ISSUE THREAD

**Issue by kouohhashi** [SeanNaren/deepspeech.pytorch#515]
HI, 

I'm training a model with SpecAugment and I encountered non loss error.
I don't know  SpecAugment is the cause but I don't experience the problem without SpecAugment.

Here's my training logs.

```
Training Summary Epoch: [1] Time taken (s): 6261  Average Loss 40.763 
100%|██████████| 808/808 [04:24<00:00,  3.06it/s]Validation Summary Epoch: [1]  Average WER 47.734  Average CER 16.047  
Training Summary Epoch: [2] Time taken (s): 7987  Average Loss 22.767 
100%|██████████| 808/808 [04:25<00:00,  3.04it/s]Validation Summary Epoch: [2]  Average WER 39.848  Average CER 11.952  
Training Summary Epoch: [3] Time taken (s): 12335 Average Loss 19.397 
100%|██████████| 808/808 [04:22<00:00,  3.08it/s]Validation Summary Epoch: [3]  Average WER 32.247  Average CER 9.342 
Training Summary Epoch: [4] Time taken (s): 13437 Average Loss 17.407 
100%|██████████| 808/808 [04:23<00:00,  3.07it/s]Validation Summary Epoch: [4]  Average WER 27.748  Average CER 7.818 
Training Summary Epoch: [5] Time taken (s): 13696 Average Loss 15.973 
100%|██████████| 808/808 [04:23<00:00,  3.07it/s]Validation Summary Epoch: [5]  Average WER 26.569  Average CER 7.085 
Training Summary Epoch: [6] Time taken (s): 13689 Average Loss 14.929 
100%|██████████| 808/808 [04:23<00:00,  3.07it/s]Validation Summary Epoch: [6]  Average WER 23.318  Average CER 6.254 
Training Summary Epoch: [7] Time taken (s): 13773 Average Loss 14.049 
100%|██████████| 808/808 [04:22<00:00,  3.08it/s]Validation Summary Epoch: [7]  Average WER 22.630  Average CER 5.813 
Training Summary Epoch: [8] Time taken (s): 13825 Average Loss 13.324 
100%|██████████| 808/808 [04:23<00:00,  3.07it/s]Validation Summary Epoch: [8]  Average WER 20.548  Average CER 5.478 
Training Summary Epoch: [9] Time taken (s): 13803 Average Loss 12.739 
100%|██████████| 808/808 [04:24<00:00,  3.06it/s]Validation Summary Epoch: [9]  Average WER 19.501  Average CER 5.075 
Training Summary Epoch: [10]  Time taken (s): 13817 Average Loss 12.281 
100%|██████████| 808/808 [04:21<00:00,  3.09it/s]Validation Summary Epoch: [10] Average WER 19.073  Average CER 4.745 
Training Summary Epoch: [11]  Time taken (s): 13841 Average Loss 11.921 
100%|██████████| 808/808 [04:23<00:00,  3.07it/s]Validation Summary Epoch: [11] Average WER 18.010  Average CER 4.657 
Training Summary Epoch: [12]  Time taken (s): 13843 Average Loss 11.510 
100%|██████████| 808/808 [04:22<00:00,  3.08it/s]Validation Summary Epoch: [12] Average WER 19.491  Average CER 4.687 
Training Summary Epoch: [13]  Time taken (s): 13877 Average Loss 11.224 
100%|██████████| 808/808 [04:23<00:00,  3.07it/s]Validation Summary Epoch: [13] Average WER 16.854  Average CER 4.230 
Training Summary Epoch: [14]  Time taken (s): 13914 Average Loss 10.971 
100%|██████████| 808/808 [04:24<00:00,  3.06it/s]Validation Summary Epoch: [14] Average WER 16.165  Average CER 4.069 
Training Summary Epoch: [15]  Time taken (s): 13937 Average Loss 10.560 
100%|██████████| 808/808 [04:24<00:00,  3.05it/s]Validation Summary Epoch: [15] Average WER 15.778  Average CER 3.905 
Training Summary Epoch: [16]  Time taken (s): 13929 Average Loss 10.355 
100%|██████████| 808/808 [04:22<00:00,  3.08it/s]Validation Summary Epoch: [16] Average WER 15.095  Average CER 3.797 
Training Summary Epoch: [17]  Time taken (s): 13961 Average Loss 10.277 
100%|██████████| 808/808 [04:25<00:00,  3.05it/s]Validation Summary Epoch: [17] Average WER 15.423  Average CER 3.754 
Training Summary Epoch: [18]  Time taken (s): 13997 Average Loss 9.985  
100%|██████████| 808/808 [04:24<00:00,  3.05it/s]Validation Summary Epoch: [18] Average WER 14.795  Average CER 3.778 
Training Summary Epoch: [19]  Time taken (s): 13986 Average Loss 9.822  
100%|██████████| 808/808 [04:23<00:00,  3.07it/s]Validation Summary Epoch: [19] Average WER 14.596  Average CER 3.689 
Training Summary Epoch: [20]  Time taken (s): 13996 Average Loss 9.593  
100%|██████████| 808/808 [04:23<00:00,  3.07it/s]Validation Summary Epoch: [20] Average WER 14.394  Average CER 3.506 
Training Summary Epoch: [21]  Time taken (s): 14003 Average Loss 9.444  
100%|██████████| 808/808 [04:23<00:00,  3.07it/s]Validation Summary Epoch: [21] Average WER 14.999  Average CER 3.575 
Training Summary Epoch: [22]  Time taken (s): 14024 Average Loss 9.330  
100%|██████████| 808/808 [04:23<00:00,  3.07it/s]Validation Summary Epoch: [22] Average WER 14.157  Average CER 3.411 
Training Summary Epoch: [23]  Time taken (s): 14052 Average Loss 9.122  
100%|██████████| 808/808 [04:23<00:00,  3.06it/s]Validation Summary Epoch: [23] Average WER 13.464  Average CER 3.451 
Training Summary Epoch: [24]  Time taken (s): 14052 Average Loss 8.988  
100%|██████████| 808/808 [04:23<00:00,  3.06it/s]Validation Summary Epoch: [24] Average WER 13.566  Average CER 3.363 
Training Summary Epoch: [25]  Time taken (s): 14017 Average Loss 8.841  
100%|██████████| 808/808 [04:25<00:00,  3.04it/s]Validation Summary Epoch: [25] Average WER 13.661  Average CER 3.251 
Training Summary Epoch: [26]  Time taken (s): 14019 Average Loss 8.823  
100%|██████████| 808/808 [04:24<00:00,  3.05it/s]Validation Summary Epoch: [26] Average WER 13.154  Average CER 3.317 
Training Summary Epoch: [27]  Time taken (s): 14060 Average Loss 8.638  
100%|██████████| 808/808 [04:23<00:00,  3.06it/s]Validation Summary Epoch: [27] Average WER 12.758  Average CER 3.136 
Training Summary Epoch: [28]  Time taken (s): 14033 Average Loss 8.534  
100%|██████████| 808/808 [04:24<00:00,  3.05it/s]Validation Summary Epoch: [28] Average WER 13.082  Average CER 3.103 
Training Summary Epoch: [29]  Time taken (s): 14003 Average Loss 8.516  
100%|██████████| 808/808 [04:25<00:00,  3.04it/s]Validation Summary Epoch: [29] Average WER 12.538  Average CER 3.108 
Training Summary Epoch: [30]  Time taken (s): 14026 Average Loss 8.424  
100%|██████████| 808/808 [04:25<00:00,  3.05it/s]Validation Summary Epoch: [30] Average WER 12.430  Average CER 3.073 
Training Summary Epoch: [31]  Time taken (s): 14062 Average Loss 8.304  
100%|██████████| 808/808 [04:26<00:00,  3.03it/s]Validation Summary Epoch: [31] Average WER 14.716  Average CER 3.429 
Training Summary Epoch: [32]  Time taken (s): 14046 Average Loss 8.249  
100%|██████████| 808/808 [04:26<00:00,  3.03it/s]Validation Summary Epoch: [32] Average WER 12.348  Average CER 3.059 
Training Summary Epoch: [33]  Time taken (s): 14016 Average Loss 8.176  
100%|██████████| 808/808 [04:24<00:00,  3.05it/s]Validation Summary Epoch: [33] Average WER 13.397  Average CER 3.111 
Training Summary Epoch: [34]  Time taken (s): 14045 Average Loss 4.346  
100%|██████████| 808/808 [04:16<00:00,  3.15it/s]Validation Summary Epoch: [34] Average WER 100.000 Average CER 100.000 
Training Summary Epoch: [35]  Time taken (s): 13955 Average Loss 0.001  
100%|██████████| 808/808 [04:14<00:00,  3.18it/s]Validation Summary Epoch: [35] Average WER 100.000 Average CER 100.000 
Training Summary Epoch: [36]  Time taken (s): 13899 Average Loss 0.000  
```

When non loss happen is not consistent.

Here's my training options.
```
python train.py --rnn-type lstm --hidden-size 1024 --hidden-layers 5 \
 --train-manifest train.csv  --val-manifest valid.csv --speed-volume-perturb \
 --spec-augment --epochs 100 --num-workers  8 --cuda  --learning-anneal 1.01 \
 --batch-size 32 --no-sortaGrad --opt-level O1 --loss-scale 1 --id my_id \
 --checkpoint --save-folder snapshots \
 --model-path snapshots/deepspeech_final.pth
```

Since a model trained without SpecAugment is likely to be overfitted, I think SpecAugment is a game changer and I want to use it.

Could someone give me some hint about how I can debug this non loss error?
Or can't I use  ```--speed-volume-perturb``` and ```--spec-augment``` both at the same time?

Thanks in advance.

**Comment by kouohhashi** [comment]
I fixed the problem by modifying ./data/sparse_image_warp.py a little bit.
Actually there were warnings like this.

```
alpha = torch.tensor(queries - floor, dtype=grid_type)
./data/sparse_image_warp.py:375: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
```

I don't know exactly when this warnings happen but changing line 375 fixed the problem. Probably this was an edge case.

From:
```
alpha = torch.tensor(queries - floor, dtype=grid_type)
```

To:
```
alpha = queries.clone().detach() - floor.clone().detach()
```
